lib/Doctrine/ORM/Query/Parser.php

```diff
@@ -311,11 +311,15 @@ public function match($token)
     {
         $lookaheadType = $this->lexer->lookahead['type'];
 
-        // short-circuit on first condition, usually types match
-        if ($lookaheadType !== $token && $token !== Lexer::T_IDENTIFIER && $lookaheadType <= Lexer::T_IDENTIFIER) {
+        /*
+         * The following condition means:
+         * - If next token matches expectation -> ok.
+         * - If expectation is to get T_IDENTIFIER and we find one of the tokens that are reserved words *but* would work as identifiers as well (e. g. "FROM", DDC-505) -> ok.
+         * - Else fail.
+         */
+        if ($lookaheadType !== $token && ($token !== Lexer::T_IDENTIFIER || $lookaheadType <= Lexer::T_IDENTIFIER)) {
```

Comment:
Isn't the second check useless? `$lookaheadType <= Lexer::T_IDENTIFIER` is always true if `$token !== Lexer::T_IDENTIFIER`